internal/controllers/clustercatalog_controller.go

```diff
@@ -45,31 +50,44 @@ func (r *ClusterCatalogReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	existingCatalog := &catalogd.ClusterCatalog{}
 	err := r.Client.Get(ctx, req.NamespacedName, existingCatalog)
 	if apierrors.IsNotFound(err) {
-		return ctrl.Result{}, r.Cache.Remove(req.Name)
+		if err := r.CatalogCache.Remove(req.Name); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error removing cache for catalog %q: %v", req.Name, err)
+		}
+		return ctrl.Result{}, nil
 	}
 	if err != nil {
 		return ctrl.Result{}, err
 	}
+
+	if existingCatalog.Status.ResolvedSource == nil ||
+		existingCatalog.Status.ResolvedSource.Image == nil ||
+		existingCatalog.Status.ResolvedSource.Image.Ref == "" {
+		// Reference is not known yet - skip cache population with no error.
+		// Once the reference is resolved another reconcile cycle
+		// will be triggered and we will progress further.
+		return ctrl.Result{}, nil
+	}
+
+	catalogFsys, err := r.CatalogCache.Get(existingCatalog.Name, existingCatalog.Status.ResolvedSource.Image.Ref)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error retrieving cache for catalog %q: %v", existingCatalog.Name, err)
```

Comment:
WDYT about propagating this error into the cache so that cache clients have more context about why the content is not available?

Reply:
@joelanford it is currently implemented how we discussed here: https://github.com/operator-framework/operator-controller/pull/1284#discussion_r1767560927

Cache client caches content fetch errors and `r.CatalogCache.Get(...)` will return fetch error to the clients. So here we will have something like `error retrieving cache for catalog "my-catalog": network request timeout or other error from actual network call`

Reply:
Ah I see now. I didn't realize most of the plumbing had already been done! Beautiful!